### PR TITLE
S3C-703: Fix werelogs design

### PR DESCRIPTION
**Pull request by DavidPineauScality**
See commit messages to have a more detailed description of the work of this branch.

As one can see, internal code changes and functional tests code changes are relatively minimal (mostly related to API + Config handling).

The Unit tests updates are bigger, because this was an opportunity to clean-up the tests relatively to the responsibility of each werelogs module.

### CODE REVIEW

**Comment by DavidPineauScality** [comment]
This should not break E2E tests; but I expect that we need to update our various components to have a proper configuration of the logging system, as this PR breaks the API for (re)configuring the logs, in order to solve a design issue,

@ironman-machine try

File: lib/api.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+const Logger = require('./Logger.js');
+
+class API {
+
+    /**
+     * This is the constructor of the Logger class. It takes optional
+     * configuration parameters, that allow to modify its behavior.
+     *
+     * @param {object} config         - A configuration object for werelogs.
+     * @param {string} config.level   - The string name of the logging level
```
Comment (michael-zapata) [inline lib/api.js]: We already have typing information in `{string}`, no need to repeat that in the doc comment.

File: lib/api.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+     *                                  ('trace', 'debug', 'info', 'warn',
+     *                                  'error' and 'fatal' in order of
+     *                                  importance.)
+     * @param {string} config.dump    - The string name of the log dumping level
```
Comment (michael-zapata) [inline lib/api.js]: Same here.

File: lib/api.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+     *                                  ('trace', 'debug', 'info', 'warn',
+     *                                  'error' and 'fatal' in order of
+     *                                  importance.)
+     * @param {object[]} config.streams - The array of streams into which to
+     *                                    log. This is an Array of objects
+     *                                    which have a field named 'stream',
+     *                                    which is writeable.
+     *
+     * @returns {undefined}
```
Comment (michael-zapata) [inline lib/api.js]: We shouldn't document this tag on the constructor.

File: lib/Logger.js [hunk 4]
```diff
@@ -11,46 +11,22 @@ class Logger {
...
-     * a more detailed description of the streams array configuration.
-     *
      * @returns {undefined}
      */
-    constructor(name, config) {
+    constructor(config, name) {
+        if (!(config instanceof Config)) {
+            throw new TypeError('Invalid parameter Type for "config".');
+        }
+        if (!(typeof name === 'string' || name instanceof String)) {
```
Comment (DavidPineauScality) [inline lib/Logger.js]: This is handled by Config, when we instanciate a Config object, or update it.

As such, this is no longer useful to do this check in Logger.js, as the responsibility moved around.

File: lib/api.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        this.config = new Config(config);
+        this.preboundLogger = Logger.bind(null, this.config);
+    }
+
+    /**
+     * This is a thunk function that allows reconfiguring the streams and log
+     * levels of all Logger and future RequestLogger objects. Note that
+     * existing RequestLogger will live their lifespan retaining the old
+     * configuration.
+     * If the provided configuration is erroneous, the function may throw
```
Comment (DavidPineauScality) [inline lib/api.js]: Took the opportunity to provide one more possible error type; and update related documentation.

File: lib/Logger.js (hunk 4 above)
Comment (michael-zapata) [inline lib/Logger.js]: Ah yes, the `throwIfInvalid()` method, got it.

**Comment by DavidPineauScality** [comment]
Merging manually since E2E succeeded for all related PRs of the issue S3C-703:

E2E launched from: https://github.com/scality/S3/pull/852#issuecomment-321001002
E2E reported successs at: https://github.com/scality/S3/pull/852#issuecomment-321015792

**Pull request merged**